src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/AttributesMustMatch.cs

```diff
@@ -0,0 +1,357 @@
+                    differences);
+
+                // parameter attributes,
+                if (leftMethod.Parameters.Length == rightMethod.Parameters.Length)
+                {
+                    for (int i = 0; i < leftMethod.Parameters.Length; i++)
+                    {
+                        ReportAttributeDifferences(
+                            left,
+                            left.GetDocumentationCommentId() + $"${i}",
+                            leftMethod.Parameters[i].GetAttributes(),
+                            rightMethod.Parameters[i].GetAttributes(),
+                            differences);
+                    }
+                }
+
+                // and type parameter attributes.
+                if (leftMethod.TypeParameters.Length == rightMethod.TypeParameters.Length)
+                {
+                    for (int i = 0; i < leftMethod.TypeParameters.Length; i++)
+                    {
+                        ReportAttributeDifferences(
+                            left,
+                            left.GetDocumentationCommentId() + $"<{i}>",
+                            leftMethod.TypeParameters[i].GetAttributes(),
+                            rightMethod.TypeParameters[i].GetAttributes(),
+                            differences);
+                    }
+                }
+            }
+
+            ReportAttributeDifferences(
+                left,
+                left.GetDocumentationCommentId() ?? "",
+                left.GetAttributes(),
+                right.GetAttributes(),
+                differences);
+        }
+
+        private class AttributeGroup
+        {
+            public readonly AttributeData Representative;
+            public readonly List<AttributeData> Attributes = new();
+            public readonly List<bool> Seen = new();
+
+            public AttributeGroup(AttributeData attr)
+            {
+                Representative = attr;
+                Seen = new List<bool>();
+                Add(attr);
+            }
+
+            public void Add(AttributeData attr)
+            {
+                Attributes.Add(attr);
+                Seen.Add(false);
+            }
+        }
+        private class AttributeSet : IEnumerable<AttributeGroup>
+        {
+            private readonly List<AttributeGroup> _set;
+            private readonly RuleSettings _settings;
+
+            public AttributeSet(RuleSettings Settings, IList<AttributeData> attributes)
+            {
+                _set = new List<AttributeGroup>();
+                _settings = Settings;
+                for (int i = 0; i < attributes.Count; i++)
+                {
+                    Add(attributes[i]);
+                }
+            }
+
+            public void Add(AttributeData attr)
+            {
+                foreach (AttributeGroup group in _set)
+                {
+                    if (_settings.SymbolComparer.Equals(group.Representative.AttributeClass!, attr.AttributeClass!))
```

Comment:
I might be wrong but I don't think you need the `!` band operator, as the EqualityComparer allows nullable parameters: https://github.com/dotnet/sdk/blob/d9d39eba26dce167c60de8b47316734311e590c0/src/ApiCompat/Microsoft.DotNet.ApiCompatibility/DefaultSymbolsEqualityComparer.cs#L13

```suggestion
                    if (_settings.SymbolComparer.Equals(group.Representative.AttributeClass, attr.AttributeClass))
```